Results page extra info (including links)

beseven opened the pull request:
<!--
Thanks for wanting to contribute to this repository.

In order for the changes to be integrated into the repo with as little friction
as possible please follow the guidance here. This includes completing all
sections as fully as possible.
Prior to creating a Pull Request, please ensure there is an open issue for the
changes you wish to make. This will provide visibility to others early in the
process. Potentially other people will wish to help out. It also allows us to
validate the change is inline with our vision for the product.

Provide a general summary of your changes in the Title
-->

## Description
Results page should offer context and alternative services for the user

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://jira.service.nhs.uk/browse/PCFC-126

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Generic results page should have bespoke extra info based on the service

## Checklist
<!-- Ensure each of the points below have been considered and completed where applicable -->
- [x] New and/or updated tests

markysoft commented on app/lib/utils/queryMapper.js:
```diff
@@ -25,39 +25,77 @@ function getLocationHeading(query) {
+  let resultsExplanation;
+  let resultsHeading;
+  let resultsOnwardsJourneyPartial;
+  if (ageChoice === constants.serviceChoices['16to25']) {
```
I'm not keen on using `let` when a `const` could be used - the logic might be clearer if the `let`s were removed and each `if` statement returned an object instead.

It would also be clearer what is returned when either of those cases are not matched, at the moment correctResultsParam is either `true` or `undefined`?

markysoft commented on app/lib/utils/queryMapper.js:
```diff
@@ -25,39 +25,77 @@ function getLocationHeading(query) {
+  let correctResultsParams;
+  let resultsExplanation;
+  let resultsHeading;
+  let resultsOnwardsJourneyPartial;
```
See the comment on `let` above

st3v3nhunt commented on test/unit/lib/utils/queryMapper.js:
```diff
@@ -49,86 +49,179 @@ describe('queryMapper', () => {
+      });
     });
 
     it('should return a sexual professionals message when the query is related to professionals and origin is any', () => {
```
The title of the test doesn't match what is being tested. If the origin could be any, perhaps have a test for each of the different origins, this should avoid confusion in the future and catch changes that affect the case where the origin isn't what is specifically being tested here.

st3v3nhunt commented on test/unit/lib/utils/queryMapper.js:
```diff
@@ -49,86 +49,179 @@ describe('queryMapper', () => {
-    });
+    describe('resultsHeading', () => {
+      const location = 'ls1';
+      it('should return a sexual professionals message when the query is related to professionals and origin is any', () => {
```
See comment above, it would be good to have each specific origin tested.

st3v3nhunt commented on test/unit/lib/utils/queryMapper.js:
```diff
@@ -49,86 +49,179 @@ describe('queryMapper', () => {
-      expect(output).to.equal('Here is a list of places where you can get tested by a sexual health professional.');
+    describe('resultsExplanation', () => {
+      const location = 'ls1';
+      it('should return a sexual professionals message when the query is related to professionals and origin is any', () => {
```
See comment above, it would be good to have each specific origin tested.

st3v3nhunt commented on test/unit/lib/utils/queryMapper.js:
```diff
@@ -49,86 +49,179 @@ describe('queryMapper', () => {
-        'them by Freepost to be tested. You\'ll usually get the results within 2 weeks.');
+    describe('correctResultsParams', () => {
+      const location = 'ls1';
+      it('should return a sexual professionals message when the query is related to professionals and origin is any', () => {
```
See comment above, it would be good to have each specific origin tested.

st3v3nhunt commented on app/lib/utils/queryMapper.js:
```diff
@@ -22,42 +22,79 @@ function getLocationHeading(query) {
-      }
-    }
+function getResultsInfoForProfessionalsByAge(ageChoice, location) {
+  const correctResultsParams = true;
```
These 3 `const`s aren't needed and could be set directly in the returned object, as some of the functions in this module do.

beseven commented on app/views/includes/onwardsJourneyKitOver25.nunjucks:
```diff
@@ -0,0 +1 @@
+<p class='link2'>Or you can <a href='{{ internalLink(siteRoot, location, 'professional', origin) }}'>see a list of places where you can get tested by a sexual health professional</a>.</p>
```
https://intellij-support.jetbrains.com/hc/en-us/community/posts/207071445-On-save-new-line-added-to-end-of-file-How-stop-this-

beseven commented on docker-compose.yml:
```diff
@@ -12,6 +12,7 @@ services:
       NODE_ENV: ${NODE_ENV-development}
       ES_HOST: ${ES_HOST}
       ES_PORT: ${ES_PORT}
+      COMPOSE_HTTP_TIMEOUT: 200
```
Tests on ES while I was on a really bad connection. I'll remove it.

beseven commented:
Addressed the issues raised. Added new tests and refactored the tests to include more specific cases and for the integration tests split it so testing the structure vs testing some content (will change with the new sources). Re: snyk, until there is a viable alternative, I would prefer having it in and solve any issues.

st3v3nhunt commented on app/middleware/prerender.js:
```diff
@@ -1,7 +1,10 @@
-  res.locals.services = mapLink.addUrl(res.locals.postcodeLocationDetails, res.locals.services);
+  res.locals.services =
+    links.addMapUrl(res.locals.postcodeLocationDetails, res.locals.services);
+  res.locals.resultsInternalLink =
```
Should this property be `resultsExternalUrl` given the function call is `addExternalUrl`?

st3v3nhunt commented on app/lib/postcodeLinks.js:
```diff
@@ -31,6 +31,14 @@ function addUrl(postcodeLocationDetails, inputList) {
   });
 }
 
+function addExternalUrl(postcodeLocationDetails, location) {
```
Rather than adding this function to this module I suggest a new module is created to keep a single responsibility.
Also, the function isn't adding a URL, it is merely returning one. Please rename to something like `getChoicesResultsUrlToOnlineTests`.

beseven replied:
Given they both generate links based on lat lon from a postcode it felt right adding them in here.  Will change the name, though.

st3v3nhunt replied:
Typically, modules are named by what they do rather than what they use to do what they do.
Also, it is common to use small, well defined modules, each doing one thing rather than bundling functions that may only be tangentially linked with one another into a module (obviously this isn't always the case and modules like the incredibly generically named `utils` are an exception to this rule but perhaps not a good thing!).
The name of this module is not going to help people in the future looking for these functions. If the functions are to exist within the same module, it should be renamed to indicate what functionality it provides. Something like `externalUrlGenerator` would do it.

st3v3nhunt commented on app/lib/postcodeLinks.js:
```diff
@@ -31,6 +31,14 @@ function addUrl(postcodeLocationDetails, inputList) {
+  if (location && postcodeLocationDetails && postcodeLocationDetails.location) {
+    return `https://www.nhs.uk/service-search/Chlamydia-free-online-tests-for-u-25s/${location}/Results/105/${postcodeLocationDetails.location.lon}/${postcodeLocationDetails.location.lat}/344/0?distance=25`;
+  }
+  return false;
```
There is no need to return false and in fact, returning different types from the same function is likely to cause confusion in the future.

beseven replied:
Kind of moved from `undefined` to `false` which are both falsey. Moving back to undefined then.

st3v3nhunt commented on app/lib/utils/queryMapper.js:
```diff
@@ -134,6 +148,8 @@ function mapServiceChoice(query) {
 module.exports = {
   getLocationHeading,
   getResultsInfo,
+  getResultsInternalLink,
```
This function is only used outside of this module for testing. Rather than exporting it on the `queryMapper` module, it would be cleaner to move it to a utility type module so it could still be tested without polluting this module's interface.

st3v3nhunt commented on app/lib/utils/queryMapper.js:
```diff
@@ -134,6 +148,8 @@ function mapServiceChoice(query) {
   getLocationHeading,
   getResultsInfo,
+  getResultsInternalLink,
+  isProfessionalChoice,
```
See the comment above.

st3v3nhunt commented on test/integration/resultsPage.js:
```diff
@@ -0,0 +1,119 @@
+        expect(searchResultsName).to.not.equal(undefined);
+        expect(searchResultsAddress).to.not.equal(undefined);
+        expect(searchResultsPhone).to.not.equal(undefined);
+        expect(searchResultsMapLink.text()).to.contain('See map and directions');
```
It looks like there isn't any additional text, does the `contain` need to be here or could it be `equal`?

beseven replied:
There's screen reader text, so contains is correct.

st3v3nhunt commented on test/integration/resultsPage.js:
```diff
@@ -0,0 +1,119 @@
+        expect(searchResultsAddress).to.not.equal(undefined);
+        expect(searchResultsPhone).to.not.equal(undefined);
+        expect(searchResultsMapLink.text()).to.contain('See map and directions');
+        expect(searchResultsOpeningTimes.text()).to.contain('See opening times');
```
See comment above.

beseven replied:
Changed to equal, with a string repeating 30 times.

st3v3nhunt commented on test/integration/resultsPage.js:
```diff
@@ -0,0 +1,119 @@
+        expect(searchResultsPhone).to.not.equal(undefined);
+        expect(searchResultsMapLink.text()).to.contain('See map and directions');
+        expect(searchResultsOpeningTimes.text()).to.contain('See opening times');
+        expect(searchResultsService.text()).to.contain('See service information');
```
See comment above.

beseven replied:
Changed to equal, with a string repeating 30 times.

st3v3nhunt commented on app/lib/postcodeLinks.js:
```diff
@@ -13,7 +13,7 @@ function filterNotFoundNames(name, address) {
+function addMapUrl(postcodeLocationDetails, inputList) {
   return inputList.map((item) => {
     const address = joinAllTruthyValues(item.address);
     const saddr = `${postcodeLocationDetails.location.lat},${postcodeLocationDetails.location.lon}`;
```
I know this hasn't changed but why is the lat/lon of the postcode being used for the start address rather than the postcode? Pharmacy finder uses the postcode as it is more obvious when clicking through to Google maps where the start address has come from. By using  the lat/lon the address is mapped to something else which might not be obvious as to how it was calculated.

beseven replied:
You can search both by outcode or by postcode - the lat long would cover both. I can raise an issue and either decompose the postcodeLocationDetails (different keys based on those) or use the users input. Which one do you suggest?

st3v3nhunt replied:
Either outcode or postcode can be used for `saddr`. I'd go with setting it based on the input. Good idea to get an issue raised to cover this.

beseven replied:
Raised https://github.com/nhsuk/sexual-health-service-finder/issues/29 to cover this

st3v3nhunt commented on app/lib/utils/utils.js:
```diff
@@ -1,14 +1,35 @@
+    && (getValues(constants.serviceChoices).includes(query.origin)));
+}
+
+function getResultsInternalLink(query, location) {
```
With the change from `link` to `url` for both `resultsInternal___` and `resultsExternal___`, can this be updated to `getResultsInternalUrl` for consistency please?
